### ISSUE TITLE
Bring back combining coverage reports from special remotes

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#4956]
Discovered in the course of https://github.com/datalad/datalad/pull/4955/files that we fail to merge coverage reports from special remotes:  https://travis-ci.org/github/datalad/datalad/jobs/729780153#L2438
```
6.32s$ if [ -z "${_DL_CRON:-}" ]; then cp __testhome__/.coverage .; coverage combine -a /tmp/.coverage-entrypoints-*; codecov; fi
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/bin/coverage", line 8, in <module>
    sys.exit(main())
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/cmdline.py", line 865, in main
    status = CoverageScript().command_line(argv)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/cmdline.py", line 588, in command_line
    self.coverage.combine(data_dirs, strict=True)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/control.py", line 696, in combine
    combine_parallel_data(self._data, aliases=aliases, data_paths=data_paths, strict=strict)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/data.py", line 117, in combine_parallel_data
    data.update(new_data, aliases=aliases)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/sqldata.py", line 706, in update
    for (file, context), numbits in lines.items()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coverage/sqldata.py", line 706, in <listcomp>
    for (file, context), numbits in lines.items()
KeyError: '/home/travis/build/datalad/datalad/datalad/api.py'
```

so that step fails and we just submit main coverage.

@jwodder - since you dealt with coverage reporting recently, could you please also have a look at this, more obscure case as well?  I have not yet tried to reproduce locally etc.  Look into .travis.yml for more pointers on how we do it (helper scripts etc)

**Comment by jwodder** [comment]
@yarikoptic What creates the `/tmp/.coverage-entrypoints-*` files?  They're not appearing in my initial tests to reproduce the issue.

**Comment by yarikoptic** [comment]
We provide cmdline shims under https://github.com/datalad/datalad/tree/master/tools/coverage-bin which is added to the PATH in `.travis.yml` while running tests.  Those shims then wrap underlying entry points through coverage (done in https://github.com/datalad/datalad/blob/master/tools/coverage-bin/datalad symlinked into other shims)

**Comment by jwodder** [comment]
@yarikoptic What tests do I have to select in order for this path to be used?  `-A "not(integration or usecase or slow or network) and not(turtle)"` doesn't produce any `/tmp/.coverage-entrypoints-*` files.

Also, what exactly is the point of doing this?  I'm pretty sure coverage can capture subprocess stats out of the box.

**Comment by yarikoptic** [comment]
> I'm pretty sure coverage can capture subprocess stats out of the box.

hm, that would have been nice but I do not think it does that.  See e.g. my comment in https://github.com/datalad/datalad/pull/4955/files#diff-e49d9c32e6cd01431c5904927c53ad0dL79 where no coverage is reported although that code is excercised by the tests within `git-annex-remote-datalad` process.  

In our case it is a sandwich of `datalad -> git-annex (non python) -> git-annex-remote-datalad` (note that stdin/stdout are used for communication so can't just gdb or even epdb there) . Testing of `datalad/customremotes` should trigger the code in customremotes sandwiched under.  Sandwich could get even deeper (`datalad -> git-annex -> git-annex-remote-datalad-archives -> git-annex -> git-annex-remote-datalad`) if underlying archive can be fetched only via datalad special remote (e.g. the case with crcns datasets).

**Comment by yarikoptic** [comment]
> doesn't produce any /tmp/.coverage-entrypoints-* files.

and that might be the underlying problem -- why they aren't produced if they should be (according to the shims I have mentioned).

**Comment by jwodder** [comment]
@yarikoptic So do you know which tests cause `git-annex-remote-datalad` to be invoked or how I could identify such tests?

**Comment by yarikoptic** [comment]
test of `datalad/customremotes/tests/test_datalad.py` would do

**Comment by jwodder** [comment]
Trying to run those tests fails with:

```
ERROR: datalad.customremotes.tests.test_datalad.test_basic_scenario_s3
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwodder/dartmouth/datalad/venv/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/jwodder/dartmouth/datalad/datalad/customremotes/tests/test_datalad.py", line 81, in test_basic_scenario_s3
    check_basic_scenario('s3://datalad-test0-versioned/3versions-allversioned.txt')
  File "/Users/jwodder/dartmouth/datalad/datalad/tests/utils.py", line 731, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/Users/jwodder/dartmouth/datalad/datalad/tests/utils.py", line 177, in _wrap_skip_if_no_network
    return func(*args, **kwargs)
  File "/Users/jwodder/dartmouth/datalad/datalad/customremotes/tests/test_datalad.py", line 26, in check_basic_scenario
    annex.init_remote(
  File "/Users/jwodder/dartmouth/datalad/datalad/support/annexrepo.py", line 1690, in init_remote
    self._run_annex_command(
  File "/Users/jwodder/dartmouth/datalad/datalad/support/annexrepo.py", line 1009, in _run_annex_command
    return run_gitcommand_on_file_list_chunks(
  File "/Users/jwodder/dartmouth/datalad/datalad/cmd.py", line 144, in run_gitcommand_on_file_list_chunks
    results.append(func(cmd, *args, **kwargs))
  File "/Users/jwodder/dartmouth/datalad/datalad/cmd.py", line 513, in run
    raise CommandError(
datalad.support.exceptions.CommandError: CommandError: 'git-annex initremote -c annex.dotfiles=true -c annex.retry=3 datalad encryption=none type=external externaltype=datalad autoenable=true' failed with exitcode 1 under /private/var/folders/l7/wrkq93d133d8zpn36fmqrq0r0000gn/T/datalad_temp_check_basic_scenariolq8rvp3k [out: 'initremote datalad
failed'] [err: 'git-annex: external special remote protocol error, unexpectedly received "" (unable to parse command)
git-annex: initremote: 1 failed']
```

and nothing is created in `/tmp`.

**Comment by yarikoptic** [comment]
try `python -m nose -s -v datalad/customremotes/tests/test_datalad.py:test_basic_scenario_local_url`

I am surprised that it didn't skip that "s3" test for you -- did it ask for some credentials for S3 and you provided them?
Is that running on OSX?

**Comment by jwodder** [comment]
@yarikoptic If I run that command as-is, it works, but there's no coverage reporting.  If I try to emulate `.travis.yml` by running:

```
cd __testhome__
PATH=$PWD/../tools/coverage-bin:$PATH \
python -m nose -s -v \
    --with-doctest \
    --with-cov --cover-package datalad \
    --logging-level=INFO \
    ../datalad/customremotes/tests/test_datalad.py:test_basic_scenario_local_url
```

it fails with the same error as above.

The S3 test did not ask me for any credentials.  Yes, I'm running these on macOS.

**Comment by yarikoptic** [comment]
ok, sorry for dropping you in the cold water of all ad-hoc scripts we have in datalad ;)

so it is the problem in the shims:
```
$> PATH=$PWD/../tools/coverage-bin:$PATH bash -x git-annex-remote-datalad              
...
+ curdir=/home/yoh/proj/datalad/datalad-master/__testhome__/../tools/coverage-bin
+ COVERAGE=python-coverage
++ which python-coverage
+ COVERAGE_RUN=' run'
```
so script exits due to `set -eu` and inability to find `python-coverage` on my Debian system - it provides python3-coverage.  If coverage is installed via `pip`, it would be simply `coverage`, but probably it should just work as `python -m coverage`.  I have initiated PR with such a change: https://github.com/datalad/datalad/pull/4959 , let's see if it just magically works (coverage reported by codecov should jump up)

**Comment by jwodder** [comment]
@yarikoptic That PR fixes my immediate problem with getting the test to run.

**Comment by yarikoptic** [comment]
fun fun.  So I would make sure that coverage on travis and locally is the most recent. And then the question is either it is something we could workaround, or it is something to report against coverage and have it fixed in it.  FWIW -- this setup used to work, so might be a regression in coverage (or might be something we broke with some code RF).  Original/travis KeyError smells like a bug in coverage.

**Comment by jwodder** [comment]
@yarikoptic The version of `coverage` on both my machine and on Travis (checked by adding `pip list` to `.travis.yml`) is 5.3, the latest version.

**Comment by jwodder** [comment]
@yarikoptic I tried editing `.travis.yml` to not use `__testhome__` on the hunch that the moving the `.coverage` file around might be causing filepath issues.  You can see the resulting builds [here](https://travis-ci.com/github/jwodder/datalad/builds/186699841).  Several builds failed on the `datalad.cmdline.tests.test_main.test_script_shims` test, while those that didn't fail (and didn't exit early due to being cron-only) all seem to combine the coverage files without error.

**Comment by jwodder** [comment]
@yarikoptic I believe I've found a workaround: https://github.com/datalad/datalad/pull/4965

**Comment by yarikoptic** [comment]
I believe fixed by #4965
